### PR TITLE
memory_opt_opcodes: constraints for MLOAD/MSTORE/MSTORE8/CALLDATALOAD

### PR DESCRIPTION
**Pull request by naure**
* Add the constraints for the memory opcodes. See [spec](https://www.notion.so/scrollzkp/Word-Addressable-Memory-Optimization-7ab2c03cb71b4fa3b8710756ce9c5fcb).
* Refactor the bus mapping witness generator.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/memory.rs [hunk 1]
```diff
@@ -203,11 +252,22 @@ impl<F: Field> ExecutionGadget<F> for MemoryGadget<F> {
...
         };
 
+        // TODO: rm.
+        println!("address: {:?}", address);
+        println!("value (LE):  {:?}", value.to_le_bytes());
+        println!("value_left:  {:?}", value_left.to_le_bytes());
+        println!("value_right: {:?}", value_right.to_le_bytes());
+
+        // TODO: updated values for MSTORE.
         self.value_left
```
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/memory.rs:220]: just confirm: for mload, value == value_pre for both left and right . for mstore/mstore8, value != value_prev ? for mstore*, need to add more buss mapping read for value_prev.

Comment (naure) [reply zkevm-circuits/src/evm_circuit/execution/memory.rs:220]: Correct.

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/memory.rs:220]: would you like to add it in this PR ? plan to merge this PR after added.

Comment (naure) [reply zkevm-circuits/src/evm_circuit/execution/memory.rs:220]: I will add this in another PR: https://github.com/scroll-tech/zkevm-circuits/pull/541

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/memory.rs:220]: ok, overall lgtm, will merge it.

**Pull request merged**